### ISSUE TITLE
Possible to use `@babel/plugin-transform-typescript` with this?

### ISSUE THREAD

**Issue by trusktr** [babel/babel-eslint#663]
I'm wondering if I can use `@babel/plugin-transform-typescript` along with `babel-eslint` instead of `typescript-eslint-parser`.

**Comment by skyrpex** [comment]
I think this is related https://github.com/babel/babel-eslint/pull/692

**Comment by hzoo** [comment]
Not sure it will completely work yet but #711 helps this process and we may need another PR like #692 to finish it up

**Comment by kaicataldo** [comment]
Yeah, there might be some work needed to make sure the AST looks correct with types. I know that in the past @JamesHenry talked about wanting to make sure the ASTs produced by this and typescript-estree matched. Given that the [TypeScript team is planning to contribute to ESLint to get the TS experience up to snuff](https://github.com/Microsoft/TypeScript/issues/29288), it would probably be a good idea so that any TypeScript plugins can be used with either.

**Comment by kaicataldo** [comment]
Relevant `typescript-estree` issue: https://github.com/JamesHenry/typescript-estree/issues/112

**Comment by JamesHenry** [comment]
The babel implementation may be an option for some folks. Part of the effort to align on ASTs is to facilitate that :)

However, I would not, that you will miss out on a whole category of linting if you go with babel-eslint for TypeScript support, because the TypeScript compiler is not involved. This means you will not be able to use rules which rely on type information, which can be very useful.

So, yes, it definitely has its place and will be a lightweight solution for folks, depending on their use-cases

**Comment by JamesHenry** [comment]
Just to double down on it - multiple solutions using shared primitives is a great situation IMO (shared primitive here being the AST)

**Comment by JamesHenry** [comment]
ICYMI, announcement following on from what Kai mentions above: https://eslint.org/blog/2019/01/future-typescript-eslint

**Comment by iFwu** [comment]
> ICYMI, announcement following on from what Kai mentions above: https://eslint.org/blog/2019/01/future-typescript-eslint

Thank you so much for your contribution! People were complaining about TSLint for a long time.

**Comment by wcandillon** [comment]
Yes that's incredible. Thank you so much. Here is my typescript with eslint setup: https://github.com/wcandillon/eslint-config-react-native-wcandillon
You can take inspiration from it.

It's amazing to see how things evolve and how fast they evolve.

**Comment by chengjianhua** [comment]
@JamesHenry @kaicataldo 

If I don't get it wrong, the new approach `typescript-eslint` which eslint support typescript still need a specific parser `@typescript-eslint/parser` for eslint configuration.

If that's it, I'm afraid of that we can't lint `.ts` files and `.js` files through a universal eslint configuration. It's impossible for us to specify different parser for different files right now, right?

**Comment by chengjianhua** [comment]
Use `babel-eslint` as the only parser,

In the view of users, we don't care about how eslint deal with it. We just know that `babel` supports parsing `.ts` files and `babel-eslint` supports `babel`, so `babel-eslint` supports `.ts` files if we enabled `@babel/plugin-typescript`.

TypeScript support to be able to be parsed by babel is the great success in the stand of users, and I can't take that I need to change my eslint configuration a lot and not be able to lint `.js` and `.ts` at the same time. If we are migrating my codes from `flow` or `es` to `typescript` in a progressive way, we need this feature.

**Comment by krousseau-cs** [comment]
Based on the response I received in the typescript-eslint/parser repo regarding support for babel proposal plugins, it seems like babel-loader support (or at least not blowing up) w/ typescript is the only option forward. They have no plans to support proposals.

https://github.com/typescript-eslint/typescript-eslint/issues/136

**Comment by JamesHenry** [comment]
Until last year, the TypeScript compiler was the way that TypeScript source code was parsed.

Now there are two ways:
- The TypeScript compiler
- `@babel/parser`

This is a fork in the road when it comes to parsing.

- `typescript-eslint` is a solution built on top of the TypeScript compiler "path"
- `babel-eslint` is a solution built on top of the babel "path"

It is great that we have more than one solution - but it is not possible to mix and match the parsing option from one and linting option from another right now (nor is there anything planned).

Please note, the TypeScript compiler is still the only way to type check your code. Babel parses and strips the type information, you still need the TypeScript compiler running if you want type checking.

**Comment by JamesHenry** [comment]
@chengjianhua As Kai explains, you are not correct. As of right now, the typescript-eslint codebase itself is a mixed JS and TS project. It is very much possible for ESLint to be configured to use different parsers for different file types.

**Comment by krousseau-cs** [comment]
@JamesHenry that makes sense. If I want to keep in babel land but still add typescript types on top of it can I use babel-eslint w/ typescript-eslint/plugin?

**Comment by JamesHenry** [comment]
That's exactly what this issue is open for :)

I'm personally not sure if there are any blockers right now, so I can't comment either way. My focus has been on typescript-eslint recently, I hope to contribute to babel-eslint more again in the future

**Comment by krousseau-cs** [comment]
Is there any way to run the typescript compiler and use babel proposals? It works w/ eslint when using babel-eslint but I still haven't found a way to run tsc when using proposals that aren't supported by typescript such as null coallescing or optional chaining.

**Comment by chengjianhua** [comment]
@JamesHenry Thanks~ I'm glad to get it that I was wrong. If it's able to configure different parsers for various extensions, it eliminated my doubts basically.

`babel-eslint` is not going to reach feature parity with `@typescript-eslint/parser`  makes sense, if there is no plan to track and match the shape of the AST generated by `@typescript-eslint/parser`, how can we lint typescript with babel-eslint? Is there a guide for this?

**Comment by armano2** [comment]
at @typescript-eslint we are doing alignment tests of AST generated by babel [`ast-alignment`](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/typescript-estree/tests/ast-alignment), you can find required plugins to produce correct AST in here [`parseWithBabelParser`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/tests/ast-alignment/parse.ts#L16)

We are also tracking difference between generated output [`fixtures-to-test`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/tests/ast-alignment/fixtures-to-test.ts), and reporting issues if we find some.

there is few major differences in generated AST `babel with estree` vs `typescript-eslint`.
- jsx (ranges are not same, as far as i know there is fix for that here)
- template literals (ranges are not same, as far as i know there is fix for that here)
- optional/abstract methods

and few nodes needs postprocessing to match AST generated by @typescript-eslint: 
[`preprocessBabylonAST`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/tests/ast-alignment/utils.ts#L84)
- StringLiteral
- NumericLiteral
- BooleanLiteral
- TSCallSignatureDeclaration
- TSConstructSignatureDeclaration
- TSFunctionType
- TSConstructorType
- TSMethodSignature
- TSTypeParameter
- ClassProperty
- TSExpressionWithTypeArguments

----------------------------

you can also check "big" alignment test project [typescript-estree-test](https://github.com/armano2/typescript-estree-test), i'm going to extend tests to cover `babel-eslint` when its going to support typescript

**Comment by PutziSan** [comment]
Hi,
I'm not sure if I understood everything, but I think at the moment the only way to use  ESLint with a mixed JS/TS codebase is with this config:

```javascript
module.exports = {
  parser: "babel-eslint",
  overrides: [
    {
      files: ["**/*.ts", "**/*.tsx"],
      parser: "@typescript-eslint/parser",
      plugins: ["@typescript-eslint"],
      // If adding a typescript-eslint version of an existing ESLint rule,
      // make sure to disable the ESLint rule here.
      rules: {/*...*/}
    }
  ],
};
```

Is this the (currently) correct way? Is it planned in the future that we can only use babel-eslint for both?

**Comment by diegoddox** [comment]
@PutziSan did work for you? I'm trying the same setup for I'm not getting all the liting on the `.tsx` files as on `jsx/js`

**Comment by PutziSan** [comment]
I work almost exclusively with TypeScript now and don't use eslint at the moment. 
The [eslint config](https://github.com/facebook/create-react-app/blob/master/packages/eslint-config-react-app/index.js#L57) of create-react-app by facebook seems to follow this approach as well and considering the popularity of this project I can imagine that it works.

**Comment by kaicataldo** [comment]
Thank you for the issue. Now that [`@babel/eslint-parser` has been released](https://babeljs.io/blog/2020/07/13/the-state-of-babel-eslint), we are making this repository read-only. If this is a change you would still like to advocate for, please reopen this in the [babel/babel monorepo](https://github.com/babel/babel).
